pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -19,6 +19,58 @@ in rec {
   # * https://nixos.org/nix/manual/#ssec-derivation
   #   Explanation about derivations in general
   mkDerivation =
+    fnOrAttrs:
+      if builtins.isFunction fnOrAttrs
+      then makeDerivationExtensible fnOrAttrs
+      else makeDerivationExtensible0 fnOrAttrs;
+
+  # Based off lib.makeExtensible, with modifications:
+  #  - lib.fix' -> lib.fix ∘ mkDerivation_; then inline fix
+  #  - convert `f` to an overlay
+  #  - inline overrideAttrs and make it positional instead of // to reduce allocs
+  makeDerivationExtensible = rattrs:
+    let
+      r = mkDerivation_ (f0:
+          let
+            f = self: super:
+              # Convert f0 to an overlay. Legacy is:
+              #   overrideAttrs (super: {})
+              # We want to introduce self. We follow the convention of overlays:
+              #   overrideAttrs (self: super: {})
+              # Which means the first parameter can be either self or super.
+              # This is surprising, but far better than the confusion that would
+              # arise from flipping an overlay's parameters in some cases.
+              let x = f0 super;
+              in
+                if builtins.isFunction x
+                then f0 self super # double call looks inefficient, but `f0 super` was a cheap thunk
```

Comment:
Also maybe point out since `self` is the first arg currying coudn't help anyways.